File: pcc/py_runtime/py/py_int_parse.py
"""Phase 4c: pcc-Python port of py_int_parse.c.

String-to-int parsing for int(str) / int(str, base). The arithmetic core
stays in py_int.c for now; this module only parses an int64 payload and
delegates canonical tagged-vs-heap construction to py_int_from_i64.
"""
from pcc.extern import extern, c_abi_export, c_ptr, c_int64, c_void
from pcc.unsafe import cstr, load_i8, store_i8, malloc, free, strlen, null, ptr_is_null


py_int_from_i64 = extern("py_int_from_i64", (c_int64,),       c_ptr)
py_int_mul      = extern("py_int_mul",      (c_ptr, c_ptr),   c_ptr)
py_int_add      = extern("py_int_add",      (c_ptr, c_ptr),   c_ptr)
py_decref       = extern("py_decref",       (c_ptr,),         c_void)
py_raise        = extern("py_raise",        (c_ptr,),         c_void)
py_exc_new      = extern("py_exc_new",      (c_int64, c_ptr), c_ptr)


def _byte_at(s, i: int) -> int:
    return load_i8(s, i) & 0xFF


def _is_space(c: int) -> int:
    if c == 32:
        return 1
    if c == 9:
        return 1
    if c == 10:
        return 1
    if c == 13:
        return 1
    return 0


def _digit_value(c: int) -> int:
    if c >= 48:
        if c <= 57:
            return c - 48
    if c >= 97:
        if c <= 122:
            return c - 97 + 10
    if c >= 65:
        if c <= 90:
            return c - 65 + 10
    return -1


def _has_prefix(s, i: int, lo: int, hi: int) -> int:
    if _byte_at(s, i) != 48:
        return 0
    c: int = _byte_at(s, i + 1)
    if c == lo:
        return 1
    if c == hi:
        return 1
    return 0


def _parse_bigint(s, start: int, base: int, negative: int):
    # Bigint fallback for decimals/values that exceed int64. Accumulates with
    # the general int ops (py_int_mul / py_int_add), which handle bignum.
    # ``start`` is the first digit position (after sign/prefix).
    base_obj = py_int_from_i64(base)
    acc = py_int_from_i64(0)
    i: int = start
    done: int = 0
    while done == 0:
        ch: int = _byte_at(s, i)
        if ch == 0:
            done = 1
        else:
            d: int = _digit_value(ch)
            if d < 0:
                done = 1
            elif d >= base:
                done = 1
            else:
                prod = py_int_mul(acc, base_obj)
                py_decref(acc)
                dobj = py_int_from_i64(d)
                acc = py_int_add(prod, dobj)
                py_decref(prod)
                py_decref(dobj)
                i = i + 1
    py_decref(base_obj)
    # Trailing: optional whitespace then NUL, matching the int64 path.
    while _is_space(_byte_at(s, i)) != 0:
        i = i + 1
    if _byte_at(s, i) != 0:
        py_decref(acc)
        return null()
    if negative != 0:
        neg_one = py_int_from_i64(-1)
        res = py_int_mul(acc, neg_one)
        py_decref(acc)
        py_decref(neg_one)
        return res
    return acc


@c_abi_export("py_int_from_cstr")
def py_int_from_cstr(s, base: int):
    # NULL input matches the C helper's parse-error contract.
    if ptr_is_null(s):
        return null()

    i: int = 0
    while _is_space(_byte_at(s, i)) != 0:
        i = i + 1

    negative: int = 0
    c: int = _byte_at(s, i)
    if c == 43 or c == 45:
        if c == 45:
            negative = 1
        i = i + 1

    if base == 0:
        if _has_prefix(s, i, 120, 88) != 0:
            base = 16
            i = i + 2
        elif _has_prefix(s, i, 98, 66) != 0:
            base = 2
            i = i + 2
        elif _has_prefix(s, i, 111, 79) != 0:
            base = 8
            i = i + 2
        elif _byte_at(s, i) == 48:
            base = 8
        else:
            base = 10
    elif base == 16:
        if _has_prefix(s, i, 120, 88) != 0:
            i = i + 2
    elif base == 2:
        if _has_prefix(s, i, 98, 66) != 0:
            i = i + 2
    elif base == 8:
        if _has_prefix(s, i, 111, 79) != 0:
            i = i + 2

    if base < 2:
        return null()
    if base > 36:
        return null()

    limit: int = 9223372036854775807
    max_div: int = limit // base
    max_rem: int = limit - max_div * base
    neg_div: int = max_div
    neg_rem: int = max_rem
    if negative != 0:
        if neg_rem == base - 1:
            neg_div = neg_div + 1
            neg_rem = 0
        else:
            neg_rem = neg_rem + 1

    digits_start: int = i
    value: int = 0
    saw_digit: int = 0
    min_exact: int = 0
    done: int = 0
    while done == 0:
        ch: int = _byte_at(s, i)
        if ch == 0:
            done = 1
        else:
            d: int = _digit_value(ch)
            if d < 0:
                done = 1
            elif d >= base:
                done = 1
            else:
                div_limit: int = max_div
                rem_limit: int = max_rem
                if negative != 0:
                    div_limit = neg_div
                    rem_limit = neg_rem
                if value > div_limit:
                    return _parse_bigint(s, digits_start, base, negative)
                if value == div_limit:
                    if d > rem_limit:
                        return _parse_bigint(s, digits_start, base, negative)
                    if negative != 0:
                        if value > max_div:
                            min_exact = 1
                        elif value == max_div:
                            if d > max_rem:
                                min_exact = 1
                saw_digit = 1
                i = i + 1
                if min_exact != 0:
                    done = 1
                else:
                    value = value * base + d

    if saw_digit == 0:
        return null()

    while _is_space(_byte_at(s, i)) != 0:
        i = i + 1
    if _byte_at(s, i) != 0:
        return null()

    if negative != 0:
        if min_exact != 0:
            min_i64: int = -9223372036854775807
            min_i64 = min_i64 - 1
            return py_int_from_i64(min_i64)
        return py_int_from_i64(0 - value)
    return py_int_from_i64(value)


def _repr_append_byte(buf, n: int, c: int, quote: int) -> int:
    # Append the CPython-style repr of source byte ``c`` into ``buf`` at
    # position ``n``, using ``quote`` (a byte value) as the active quote so only
    # the active quote is backslash-escaped. Mirrors repr_append_byte in
    # py_int_parse.c. ``buf`` must have room for the worst case (4 bytes).
    if c == 92 or c == quote:  # '\\' or the active quote
        store_i8(buf, n, 92)
        store_i8(buf, n + 1, c)
        return n + 2
    if c == 10:  # '\n'
        store_i8(buf, n, 92)
        store_i8(buf, n + 1, 110)  # 'n'
        return n + 2
    if c == 13:  # '\r'
        store_i8(buf, n, 92)
        store_i8(buf, n + 1, 114)  # 'r'
        return n + 2
    if c == 9:  # '\t'
        store_i8(buf, n, 92)
        store_i8(buf, n + 1, 116)  # 't'
        return n + 2
    if c < 32 or c == 127:
        store_i8(buf, n, 92)       # '\\'
        store_i8(buf, n + 1, 120)  # 'x'
        hi: int = (c >> 4) & 15
        lo: int = c & 15
        if hi < 10:
            store_i8(buf, n + 2, 48 + hi)
        else:
            store_i8(buf, n + 2, 87 + hi)  # 'a'(97) - 10
        if lo < 10:
            store_i8(buf, n + 3, 48 + lo)
        else:
            store_i8(buf, n + 3, 87 + lo)
        return n + 4
    store_i8(buf, n, c)
    return n + 1


def _build_bad_literal_message(s, base: int):
    # Build "invalid literal for int() with base <base>: <repr(s)>" into a
    # freshly malloc'd NUL-terminated buffer; caller frees. ``base`` is the
    # ORIGINAL base argument (0 renders as "base 0"). Mirrors
    # build_bad_literal_message in py_int_parse.c. Returns null() on OOM.
    slen: int = strlen(s)
    # CPython quote selection: single quote unless the string has a single
    # quote but no double quote.
    quote: int = 39  # '\''
    has_single: int = 0
    has_double: int = 0
    i: int = 0
    while i < slen:
        b: int = _byte_at(s, i)
        if b == 39:
            has_single = 1
        elif b == 34:
            has_double = 1
        i = i + 1
    if has_single != 0:
        if has_double == 0:
            quote = 34  # '"'

    # Prefix: "invalid literal for int() with base " (36 bytes).
    prefix_len: int = 36
    cap: int = prefix_len + 24 + 2 + slen * 4 + 1
    buf = malloc(cap)
    if ptr_is_null(buf):
        return null()
    prefix = cstr("invalid literal for int() with base ")
    n: int = 0
    while n < prefix_len:
        store_i8(buf, n, _byte_at(prefix, n))
        n = n + 1

    # Render base as signed decimal (handles 0 and defensively negatives).
    neg: int = 0
    ub: int = base
    if base < 0:
        neg = 1
        ub = 0 - base
    # Collect decimal digits (reverse), then append forward.
    digit_buf = malloc(24)
    d: int = 0
    if ub == 0:
        store_i8(digit_buf, 0, 48)  # '0'
        d = 1
    else:
        while ub != 0:
            store_i8(digit_buf, d, 48 + (ub % 10))
            ub = ub // 10
            d = d + 1
    if neg != 0:
        store_i8(buf, n, 45)  # '-'
        n = n + 1
    while d > 0:
        d = d - 1
        store_i8(buf, n, _byte_at(digit_buf, d))
        n = n + 1
    free(digit_buf)

    store_i8(buf, n, 58)      # ':'
    n = n + 1
    store_i8(buf, n, 32)      # ' '
    n = n + 1
    store_i8(buf, n, quote)
    n = n + 1
    i = 0
    while i < slen:
        n = _repr_append_byte(buf, n, _byte_at(s, i), quote)
        i = i + 1
    store_i8(buf, n, quote)
    n = n + 1
    store_i8(buf, n, 0)       # NUL
    return buf


@c_abi_export("py_int_from_cstr_or_raise")
def py_int_from_cstr_or_raise(s, base: int):
    # int(str) builtin: parse like py_int_from_cstr but raise ValueError on
    # invalid input instead of returning NULL (which the frontend would unbox to
    # 0 -> int('xyz') silently became 0). Mirrors py_int_from_cstr_or_raise in
    # py_int_parse.c; py_int_from_cstr stays NULL-returning for other callers.
    #
    # CPython raises two distinct ValueError messages, reproduced here:
    #   - bad base (not 0 and outside 2..36):
    #       "int() base must be >= 2 and <= 36, or 0"
    #   - unparseable literal (valid base):
    #       "invalid literal for int() with base <base>: <repr(s)>"
    # ``base`` is the ORIGINAL argument (0 stays "base 0"); the repr is a
    # CPython-accurate repr of the whole original string.
    if base != 0:
        if base < 2 or base > 36:
            py_raise(py_exc_new(2, cstr("int() base must be >= 2 and <= 36, or 0")))
            return null()
    v = py_int_from_cstr(s, base)
    if ptr_is_null(v) == 0:
        return v
    src = s
    if ptr_is_null(s):
        src = cstr("")
    msg = _build_bad_literal_message(src, base)
    if ptr_is_null(msg) == 0:
        py_raise(py_exc_new(2, msg))  # PY_EXC_VALUEERROR
        free(msg)
    else:
        py_raise(py_exc_new(2, cstr("invalid literal for int()")))
    return null()

File: pcc/py_runtime/src/py_int_parse.c
/* pcc/py_runtime/src/py_int_parse.c
 *
 * Small string-to-int helper split from py_int.c so Phase 4c can replace
 * it with the pcc-Python port in py_int_parse.py.
 */

#include "py_internal.h"
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

static int is_parse_space(unsigned char c) {
    return c == ' ' || c == '\t' || c == '\n' || c == '\r';
}

static int digit_value(unsigned char c) {
    if (c >= '0' && c <= '9') return (int)(c - '0');
    if (c >= 'a' && c <= 'z') return (int)(c - 'a') + 10;
    if (c >= 'A' && c <= 'Z') return (int)(c - 'A') + 10;
    return -1;
}

static int has_prefix(const char *s, int64_t i, unsigned char lo, unsigned char hi) {
    return s[i] == '0' && (s[i + 1] == (char)lo || s[i + 1] == (char)hi);
}

/* Bigint fallback for values exceeding int64; accumulates with the general
 * int ops (which handle bignum).  ``start`` is the first digit position. */
static PyObject *parse_bigint_cstr(const char *s, int64_t start, int base, int negative) {
    PyObject *base_obj = py_int_from_i64(base);
    PyObject *acc = py_int_from_i64(0);
    int64_t i = start;
    while (s[i] != '\0') {
        int d = digit_value((unsigned char)s[i]);
        if (d < 0 || d >= base) break;
        PyObject *prod = py_int_mul(acc, base_obj);
        py_decref(acc);
        PyObject *dobj = py_int_from_i64(d);
        acc = py_int_add(prod, dobj);
        py_decref(prod);
        py_decref(dobj);
        i++;
    }
    py_decref(base_obj);
    while (is_parse_space((unsigned char)s[i])) i++;
    if (s[i] != '\0') {
        py_decref(acc);
        return NULL;
    }
    if (negative) {
        PyObject *neg_one = py_int_from_i64(-1);
        PyObject *res = py_int_mul(acc, neg_one);
        py_decref(acc);
        py_decref(neg_one);
        return res;
    }
    return acc;
}

PyObject *py_int_from_cstr(const char *s, int base) {
    if (s == NULL) return NULL;
    int64_t i = 0;
    while (is_parse_space((unsigned char)s[i])) i++;

    int negative = 0;
    if (s[i] == '+' || s[i] == '-') {
        negative = s[i] == '-';
        i++;
    }

    if (base == 0) {
        if (has_prefix(s, i, 'x', 'X')) {
            base = 16;
            i += 2;
        } else if (has_prefix(s, i, 'b', 'B')) {
            base = 2;
            i += 2;
        } else if (has_prefix(s, i, 'o', 'O')) {
            base = 8;
            i += 2;
        } else if (s[i] == '0') {
            base = 8;
        } else {
            base = 10;
        }
    } else if (base == 16 && has_prefix(s, i, 'x', 'X')) {
        i += 2;
    } else if (base == 2 && has_prefix(s, i, 'b', 'B')) {
        i += 2;
    } else if (base == 8 && has_prefix(s, i, 'o', 'O')) {
        i += 2;
    }

    if (base < 2 || base > 36) return NULL;

    uint64_t limit = negative ? ((uint64_t)INT64_MAX + 1ULL) : (uint64_t)INT64_MAX;
    uint64_t value = 0;
    int saw_digit = 0;
    int64_t digits_start = i;
    while (s[i] != '\0') {
        int d = digit_value((unsigned char)s[i]);
        if (d < 0 || d >= base) break;
        uint64_t ud = (uint64_t)d;
        if (value > (limit - ud) / (uint64_t)base) {
            return parse_bigint_cstr(s, digits_start, base, negative);
        }
        value = value * (uint64_t)base + ud;
        saw_digit = 1;
        i++;
    }
    if (!saw_digit) return NULL;

    while (is_parse_space((unsigned char)s[i])) i++;
    if (s[i] != '\0') return NULL;

    if (negative) {
        if (value == ((uint64_t)INT64_MAX + 1ULL)) return py_int_from_i64(INT64_MIN);
        return py_int_from_i64(-(int64_t)value);
    }
    return py_int_from_i64((int64_t)value);
}

/* Append the CPython-style repr of the source byte ``c`` into ``buf`` at
 * position ``n``, using ``quote`` as the active quote char (so only the active
 * quote is backslash-escaped, matching CPython's quote-selection). ``buf`` must
 * have room for the worst case (4 bytes: ``\xHH``). Returns the new length. */
static int64_t repr_append_byte(char *buf, int64_t n, unsigned char c, char quote) {
    static const char hexdigits[] = "0123456789abcdef";
    if (c == '\\' || (char)c == quote) {
        buf[n++] = '\\';
        buf[n++] = (char)c;
    } else if (c == '\n') {
        buf[n++] = '\\';
        buf[n++] = 'n';
    } else if (c == '\r') {
        buf[n++] = '\\';
        buf[n++] = 'r';
    } else if (c == '\t') {
        buf[n++] = '\\';
        buf[n++] = 't';
    } else if (c < 0x20 || c == 0x7f) {
        buf[n++] = '\\';
        buf[n++] = 'x';
        buf[n++] = hexdigits[(c >> 4) & 0xf];
        buf[n++] = hexdigits[c & 0xf];
    } else {
        /* Printable ASCII and (byte-oriented) high bytes pass through, matching
         * CPython's default repr which only escapes non-printables. */
        buf[n++] = (char)c;
    }
    return n;
}

/* Build "invalid literal for int() with base <base>: <repr(s)>" into a freshly
 * malloc'd NUL-terminated buffer; caller frees. ``base`` is the ORIGINAL base
 * argument (0 renders as "base 0"), matching CPython's message. Returns NULL on
 * OOM. */
static char *build_bad_literal_message(const char *s, int base) {
    int64_t slen = (int64_t)strlen(s);
    /* Choose the quote char the way CPython does: single quote unless the
     * string contains a single quote but no double quote. */
    char quote = '\'';
    int has_single = 0;
    int has_double = 0;
    for (int64_t i = 0; i < slen; i++) {
        if (s[i] == '\'') has_single = 1;
        else if (s[i] == '"') has_double = 1;
    }
    if (has_single && !has_double) quote = '"';

    const char *prefix = "invalid literal for int() with base ";
    /* Worst case: each source byte expands to 4 chars (\\xHH). */
    int64_t cap = (int64_t)strlen(prefix) + 24 /* base digits + ": " */
                  + 2 /* quotes */ + slen * 4 + 1;
    char *buf = (char *)malloc((size_t)cap);
    if (buf == NULL) return NULL;
    int64_t n = 0;
    for (const char *p = prefix; *p != '\0'; p++) buf[n++] = *p;

    /* Render base as a signed decimal (handles 0 and, defensively, any value
     * that reached here). */
    char digits[24];
    int64_t d = 0;
    int b = base;
    int neg = 0;
    if (b < 0) { neg = 1; }
    unsigned int ub = neg ? (unsigned int)(-(long)b) : (unsigned int)b;
    if (ub == 0) {
        digits[d++] = '0';
    } else {
        while (ub != 0) { digits[d++] = (char)('0' + (ub % 10)); ub /= 10; }
    }
    if (neg) buf[n++] = '-';
    while (d > 0) buf[n++] = digits[--d];

    buf[n++] = ':';
    buf[n++] = ' ';
    buf[n++] = quote;
    for (int64_t i = 0; i < slen; i++) {
        n = repr_append_byte(buf, n, (unsigned char)s[i], quote);
    }
    buf[n++] = quote;
    buf[n] = '\0';
    return buf;
}

/* int(str) builtin: parse like py_int_from_cstr but raise ValueError on invalid
 * input instead of returning NULL (which the frontend would otherwise unbox to
 * 0 -> int('xyz') silently became 0). py_int_from_cstr stays NULL-returning for
 * other callers.
 *
 * CPython raises two distinct ValueError messages here, which this reproduces:
 *   - a bad base (not 0 and outside 2..36):
 *       "int() base must be >= 2 and <= 36, or 0"
 *   - an unparseable literal (valid base):
 *       "invalid literal for int() with base <base>: <repr(s)>"
 * ``base`` is the ORIGINAL argument (0 stays "base 0"; the resolved base is not
 * shown), and the repr is a CPython-accurate repr of the whole original string
 * (whitespace and all), with CPython quote selection and \\x/\\n/\\r/\\t escapes. */
/* int(o[, base]) with an OBJECT result.
 *
 * The frontend's `int(<dyn>)` lowering unboxes every branch to i64 and phis
 * them, which truncates a bignum to 0.  Callers that want the object
 * projection emit ONE call to this instead, so no basic blocks are created in
 * the frontend probe that asks for it.
 *
 * Returns a NEW reference, or NULL with a pending exception.  `base` applies
 * only to the string case, matching the existing lowering.
 */
PyObject *py_obj_as_int_object(PyObject *o, int base) {
    if (o == NULL) return NULL;
    int64_t tag = py_obj_type_tag(o);
    if (tag == PY_TYPE_STR) {
        return py_int_from_cstr_or_raise(py_str_utf8(o), base);
    }
    if (tag == PY_TYPE_FLOAT) {
        return py_int_from_i64((int64_t)py_float_to_f64(o));
    }
    if (tag == PY_TYPE_BOOL) {
        return py_int_from_i64(py_obj_truthy(o));
    }
    py_incref(o);
    return o;
}

PyObject *py_int_from_cstr_or_raise(const char *s, int base) {
    if (base != 0 && (base < 2 || base > 36)) {
        py_raise(py_exc_new(PY_EXC_VALUEERROR,
                            "int() base must be >= 2 and <= 36, or 0"));
        return NULL;
    }
    PyObject *v = py_int_from_cstr(s, base);
    if (v == NULL) {
        const char *src = (s != NULL) ? s : "";
        char *msg = build_bad_literal_message(src, base);
        if (msg != NULL) {
            py_raise(py_exc_new(PY_EXC_VALUEERROR, msg));
            free(msg);
        } else {
            py_raise(py_exc_new(PY_EXC_VALUEERROR,
                                "invalid literal for int()"));
        }
    }
    return v;
}
